src/engine/subscription/Subscription.spec.ts

```diff
@@ -0,0 +1,575 @@
+          ],
+        });
+
+        otherEntity.getSubscriptionByName('onBuild');
+      }
+
+      expect(fn).toThrowErrorMatchingSnapshot();
+    });
+
+    it('should throw on duplicate subscription names', () => {
+      const subscriptions = [
+        new Subscription({
+          type: SUBSCRIPTION_TYPE_CREATE,
+          name: 'onBuild',
+          description: 'build item',
+          attributes: ['someAttribute'],
+        }),
+        new Subscription({
+          type: SUBSCRIPTION_TYPE_DELETE,
+          name: 'onBuild',
+          description: 'build item',
+          attributes: ['someAttribute'],
+        }),
+      ];
+
+      function fn() {
+        processEntitySubscriptions(entity, subscriptions);
+      }
+
+      expect(fn).toThrowErrorMatchingSnapshot();
+    });
+
+    it('should throw if unknown attributes are used', () => {
+      const subscriptions = [
+        new Subscription({
+          type: SUBSCRIPTION_TYPE_CREATE,
+          name: 'onBuild',
+          description: 'build item',
+          attributes: ['doesNotExist'],
+        }),
+      ];
+
+      function fn() {
+        processEntitySubscriptions(entity, subscriptions);
+      }
+
+      expect(fn).toThrowErrorMatchingSnapshot();
+    });
+
+    it('should allow for empty attribute lists on DELETE type subscriptions', () => {
+      const subscriptions = [
+        new Subscription({
+          type: SUBSCRIPTION_TYPE_DELETE,
+          name: 'onDrop',
+          description: 'drop item',
+          attributes: [],
+        }),
+      ];
+
+      processEntitySubscriptions(entity, subscriptions);
+    });
+  });
+
+  describe('preProcessor', () => {
+    const testEntity = new Entity({
+      name: 'SomeTestsEntityName',
+      description: 'Just some description',
+      attributes: {
+        someAttribute: {
+          type: DataTypeString,
+          description: 'Just some description',
+          required: false,
+        },
+        anotherAttribute: {
+          type: DataTypeString,
+          description: 'Just some description',
+        },
+      },
+      subscriptions: [
+        new Subscription({
+          name: 'SomeSubWithPreProcessor',
+          type: SUBSCRIPTION_TYPE_CREATE,
+          description: 'build item',
+          attributes: ['someAttribute'],
+          delimiter: '/',
+          // wildCard: '',
+          // pattern: '',
+          preProcessor: (
+            _entity,
+            _id,
+            _source,
+            input,
+            typeName,
+            entitySubscription,
+          ) => {
+            if (entitySubscription.attributes && Object.keys(input).length) {
+              const delimiter = entitySubscription.delimiter;
+              const filled = entitySubscription.attributes
+                .map(attribute => input[attribute])
+                .reduce((acc, curr) => `${acc + delimiter + curr}`, '');
+
+              const topic = `${entitySubscription.name}${_entity.name}${filled}`;
```

Comment:
so the attributes are part of the topic name.
can you pls elaborate on that

Reply:
With publish-subscribe pattern, subscribers receive every messages published to the topics on which they subscribed. 

Let's take an example for a Car model with id and ownerId attributes.
After we create/update an instance, we could publish a message on topic `onCreateCar/${ownerId}/${id}` or `onUpdateCar/${ownerId}/${id}` 
- If you subscribe to `onCreateCar/#` (# being a wildcard acting like a catchall for every messages topic beginning by `onCreateCar/`), you would receive a message for each Car instance created
- If you subscribe to `onCreateCar/${ownerId}` you would receive a message for each Car instance belonging to ${ownerId}
- And so on...

And of course accessControl needs to be done so any user is not allowed to subscribe to any topic and since the Permission system in Shyft is partially based on the Entity attributes i thought it would make sense to use these attributes to compose the topic chain.

You can see an example in action in Entourage-server, where we can subscribe to `profileDestroyed/${profile}/${version}` to only receive update from a specific profile version.

Reply:
I see, cool 👍